src/main/java/com/oracle/nosql/spring/data/repository/support/NosqlEntityInformation.java

```diff
@@ -616,4 +461,157 @@ public Map<String, FieldValue.Type> getShardKeys() {
+                                                idFieldClass)
+                                );
+                            }
+                            ss.add(primaryKey.getName());
+                            nonShardMap.put(order, ss);
+                        }
+                    }
+                }
+
+                List<String> sortedShardKeys = new ArrayList<>();
+                List<String> sortedNonShardKeys = new ArrayList<>();
+
+                shardMap.forEach((order, keys) -> {
+                    //order should be specified on all fields if at all is used
+                    if (order != NOTSET_PRIMARY_KEY_ORDER &&
+                            shardMap.get(NOTSET_PRIMARY_KEY_ORDER) != null) {
+                        throw new IllegalArgumentException("If order is " +
+                                "specified, it must be specified on all key" +
+                                " fields of the composite key class " +
+                                idField.getType().getName());
+                    }
+
+                    //order value must be unique
+                    if (order != NOTSET_PRIMARY_KEY_ORDER && keys.size() > 1) {
+                        throw new IllegalArgumentException("Order of " +
+                                "keys must be unique in composite key " +
+                                "class " + idFieldClass.getName());
+                    }
+                    sortedShardKeys.addAll(keys);
+                });
+
+                nonShardMap.forEach((order, keys) -> {
+                    //order should be specified on all fields if at all is used
+                    if (order != NOTSET_PRIMARY_KEY_ORDER &&
+                            nonShardMap.get(NOTSET_PRIMARY_KEY_ORDER) != null) {
+                        throw new IllegalArgumentException("If order is " +
+                                "specified, it must be specified on all key" +
+                                " fields of the composite key class " +
+                                idField.getType().getName());
+                    }
+
+                    //order value must be unique
+                    if (order != NOTSET_PRIMARY_KEY_ORDER && keys.size() > 1) {
+                        throw new IllegalArgumentException("Order of " +
+                                "keys must be unique in composite key " +
+                                "class " + idFieldClass.getName());
+                    }
+                    sortedNonShardKeys.addAll(keys);
+                });
+
+                if (sortedShardKeys.isEmpty()) {
+                    throw new IllegalArgumentException("At least one of the " +
+                            "@NosqlKey must be shard key in class " +
+                            idFieldClass.getName());
+                }
+
+                if (!nonShardMap.isEmpty()) {
+                    int shardMaxOrder =
+                            (int) ((TreeMap<?, ?>) shardMap).lastKey();
```

Comment:
To avoid the up castings I would declare the shardMap and nonShardMap as TreeMap.